### PR TITLE
[VEN-2262]: Functionality for seizing xvs tokens

### PR DESCRIPTION
**Pull request by Debugger022**
## Description

Resolves #VEN-2262

This PR contains functionality for seizing the **XVS** tokens.

 ## Checklist

- [ ] I have updated the documentation to account for the changes in the code.
- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a test preventing this bug from silently reappearing again.
- [ ] My contribution follows [Venus contribution guidelines](docs/CONTRIBUTING.md).

### CODE REVIEW

File: contracts/Comptroller/Diamond/facets/RewardFacet.sol [hunk 1]
```diff
@@ -134,6 +134,32 @@ contract RewardFacet is IRewardFacet, XVSRewardsHelper {
...
+        ensureAllowed("seizeVenus(address[],address)");
+
+        uint256 holdersLength = holders.length;
+        uint256 totalHoldings;
+
+        updateAndDistributeRewards(holders, allMarkets, true, true);
+        for (uint256 j; j < holdersLength; ++j) {
+            address holder = holders[j];
+            totalHoldings += venusAccrued[holder];
+            delete venusAccrued[holder];
```
Comment (chechu) [inline contracts/Comptroller/Diamond/facets/RewardFacet.sol:152]: Should we emit an event with the seized amount and the holder, per holder?

File: contracts/Comptroller/Diamond/facets/RewardFacet.sol [hunk 2]
```diff
@@ -151,7 +177,44 @@ contract RewardFacet is IRewardFacet, XVSRewardsHelper {
...
+    /**
+     * @notice Update and distribute tokens
+     * @param holders The addresses to claim XVS for
+     * @param vTokens The list of markets to claim XVS in
+     * @param borrowers Whether or not to claim XVS earned by borrowing
+     * @param suppliers Whether or not to claim XVS earned by supplying
+     */
+    function updateAndDistributeRewards(
+        address[] memory holders,
+        VToken[] memory vTokens,
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/Comptroller/Diamond/facets/RewardFacet.sol:210]: We can not use calldata here as these params are coming from `ClaimVenus` function which is a public function and having memory variables. Also we can not change `ClaimVenus` function params to calldata.

File: contracts/Comptroller/Diamond/facets/RewardFacet.sol (hunk 1 above)
Comment (Debugger022) [reply contracts/Comptroller/Diamond/facets/RewardFacet.sol:157]: Added https://github.com/VenusProtocol/venus-protocol/pull/417/commits/a63c9c9a72474a5f8aa1f4f79b1bc13514e830cf

File: contracts/Comptroller/Diamond/facets/RewardFacet.sol [hunk 3]
```diff
@@ -134,6 +134,32 @@ contract RewardFacet is IRewardFacet, XVSRewardsHelper {
...
+        uint256 totalHoldings;
+
+        updateAndDistributeRewards(holders, allMarkets, true, true);
+        for (uint256 j; j < holdersLength; ++j) {
+            address holder = holders[j];
+            totalHoldings += venusAccrued[holder];
+            delete venusAccrued[holder];
+        }
+
+        if (totalHoldings == 0 || totalHoldings > IBEP20(getXVSAddress()).balanceOf(address(this))) {
```
[the remark this answers is not in the document]
Comment (Debugger022) [reply contracts/Comptroller/Diamond/facets/RewardFacet.sol]: Done https://github.com/VenusProtocol/venus-protocol/pull/417/commits/a63c9c9a72474a5f8aa1f4f79b1bc13514e830cf

**Review by kkirka — COMMENTED** [review]
Would be nice to have more tests for this feature since it touches some nontrivial but quite critical bits of code, other than that LGTM

File: tests/hardhat/Fork/claimRewards.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+        await xvs.connect(xvsHotWallet).transfer(comptroller.address, parseUnits("10", 18));
+
+        const txn = await comptroller.connect(user).seizeVenus([user.address], recipient);
+
+        await expect(txn).to.emit(comptroller, "VenusSeized").withArgs(user.address, 75109993761328);
+        await expect(txn).to.emit(comptroller, "VenusGranted").withArgs(recipient, 75109993761328);
+      });
+
+      it("Fails when the XVS doesn't have enough funds", async () => {
```
Comment (chechu) [inline tests/hardhat/Fork/claimRewards.ts:145]:
```suggestion
      it("Fails when the user doesn't have enough XVS funds allocated", async () => {
```

**Review by GitGuru7 — APPROVED** [review]

**Pull request merged**